### PR TITLE
feat: add <ConcealableTabs />

### PR DESCRIPTION
**Pull request by UncleBill**
closes https://mask.atlassian.net/browse/MF-329

## Use the tail prop

```tsx
                    <ConcealableTabs<string>
                        tabs={tabs}
                        selectedId={selectedTabId}
                        onChange={setSelectedTab}
                        tail={
                            <Box display="flex" justifyContent="center" alignItems="center" height="35px" width="35px">
                                <GearSettingsIcon />
                            </Box>
                        }
                    />
```
image

## Scrolling detect

image

image

## The controllers can auto update

image

### CODE REVIEW

File: packages/mask/src/components/InjectedComponents/PageTab.tsx [hunk 1]
```diff
@@ -1,48 +1,75 @@
...
     },
 }))
 
 export interface PageTabProps {
-    tabs: ({ pluginID: string } & Plugin.SNSAdaptor.ProfileTab)[]
-    selectedTab?: Plugin.SNSAdaptor.ProfileTab
-    onChange?: (tag: Plugin.SNSAdaptor.ProfileTab) => void
+    tabs: TabOption[]
+    selectedTab?: TabOption['id']
+    onChange?: (tag: TabOption['id']) => void
```
Comment (septs) [inline packages/mask/src/components/InjectedComponents/PageTab.tsx]:
```suggestion
    onChange?(tabId: TabOption['id']): void
```

File: packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx [hunk 2]
```diff
@@ -63,31 +63,44 @@ export function ProfileTabContent(props: ProfileTabContentProps) {
...
-    const selectedTabComputed = selectedTab ?? first(tabs)
+    const availablePlugins = useAvailablePlugins(activatedPlugins)
+    const displayPlugins = useMemo(() => {
+        return availablePlugins
+            .flatMap((x) => x.ProfileTabs?.map((y) => ({ ...y, pluginID: x.ID })) ?? [])
+            .filter((z) => z.Utils?.shouldDisplay?.(identity, addressNames) ?? true)
+    }, [availablePlugins, identity, addressNames])
+    const tabs = useMemo(() => {
+        return displayPlugins
+            .sort((a, z) => {
```
Comment (septs) [inline packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx:73]: extract the comparator as function?

```ts
displayPlugins
    .sort(makeComparator(PluginId.NextID, PluginId.Collectible, /* ..., etc */))

function makeComparator(...pluginIds: PluginId[]) {
    return (a, b) => /* comparator logic */
}
```

Comment (UncleBill) [reply packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx:73]: Out of context of this PR. I would reject. You are free to refactor in another PR.

File: packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx [hunk 3]
```diff
@@ -31,9 +32,7 @@ function getTabContent(tabId: string) {
 
 const useStyles = makeStyles()((theme) => ({
     root: {},
-    tags: {
-        padding: theme.spacing(2),
-    },
+    tags: {},
     metadata: {},
```
Comment (nuanyang233) [inline packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx]: remove useless classes

Comment (UncleBill) [reply packages/mask/src/components/InjectedComponents/ProfileTabContent.tsx]: Removed except `root`

File: packages/shared/src/UI/components/ConcealableTabs/index.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    const [reachedLeftEdge, setReachedLeftEdge] = useState(false)
+    const [reachedRightEdge, setReachedRightEdge] = useState(false)
+
+    useLayoutEffect(() => {
+        const tabList = trackRef.current
+        if (!tabList) return
+        const isWider = tabList.scrollWidth > tabList.offsetWidth
+        setOverflow(isWider)
+
+        const detectScrollStatus = () => {
```
Comment (guanbinrui) [inline packages/shared/src/UI/components/ConcealableTabs/index.tsx]: Throttle it?

Comment (UncleBill) [reply packages/shared/src/UI/components/ConcealableTabs/index.tsx]: Good point, thanks.

**Review by guanbinrui — DISMISSED** [review]

**Review by guanbinrui — DISMISSED** [review]

**Comment by UncleBill** [comment]
@DimensionDev/qa Please help to verify.

File: packages/shared/src/UI/components/ConcealableTabs/index.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        tabList.addEventListener('scroll', detectScrollStatus)
+        return () => {
+            tabList.removeEventListener('scroll', detectScrollStatus)
+        }
+    }, [])
+
+    useEffect(() => {
+        if (selectedId === undefined && tabs.length) {
+            onChange?.(tabs[0].id)
+        }
```
Comment (septs) [inline packages/shared/src/UI/components/ConcealableTabs/index.tsx:154]:
```suggestion
        if (!selectedId || !tabs?.[0]?.id) return
        onChange?.(tabs[0].id)
```

File: packages/shared/src/UI/components/ConcealableTabs/index.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    }, [selectedId === undefined, tabs.length > 0, onChange])
+
+    const slide = useCallback((toLeft: boolean) => {
```
Comment (septs) [inline packages/shared/src/UI/components/ConcealableTabs/index.tsx:157]:
```suggestion
    const setSlide = useCallback((toLeft: boolean) => {
```

**Review by lelenei — APPROVED** [review]

**Review by hanszhoou — APPROVED** [review]

**Pull request merged**